### DISCUSSION TITLE
AWS SSO Connect to Identity Center(OIdC) with CyberDuck

### DISCUSSION THREAD

**Discussion by arstrandsr** [iterate-ch/cyberduck#16023] [Q&A]
I have a working OIdC connection via AWS CLI to AWS SSO (Identity Center).  Now I need to give the "credentials" to Cyberduck.  The Credentials are in the `.aws/config` file as shown below.  

I am looking for ideas/suggestions.

### Background
AWS, and I believe others are moving to Open Identity Connector SSO.    AWS has named the SSO offering _**Identity Center**_.   Identity center uses short term STS tokens exclusively.

### The goal
Connect Cyberduck to my  AWS S3 bucket.

### Tried steps

- I have searched the Cyberduck website and found pages that state OIdC is now present
- I noticed that there is an Azure and Google S3 connector
- I have looked at the the S3 profiles available from the "more options" list.

None of the options seem to look anything like the required AWS connection sequence.  The below `.aws/config` file works with AWS CLI and automatically refreshes  tokens.  

### Idea Requests

1. Ideally I give this data to Cyberduck and it works as expected.  
2. If that isn't possible, then I will write a small Python script to somehow maintain tokens and allow Cyberduck to run.
3. Any ideas that will get me closer will be appreciated.

### .aws/config file
Items like `<S3Bucket>` were manually edited here for security.

```
[profile  <S3Bucket>]
sso_session = <S3Bucketsession>
sso_account_id = <sso-account>
sso_role_name = <S3-sts-Access>
region = us-west-2
output = json

[sso-session <S3Bucketsession>]
sso_start_url = https://<AccountAlias>.awsapps.com/start/#
sso_region = us-west-2
sso_registration_scopes = sso:account:access
```

**Comment by dkocher** [comment]
In #13377. The problem is that AWS SSO (Identity Center) is [not OIDC compliant](https://github.com/iterate-ch/cyberduck/issues/13377#issuecomment-1695961553) and therefore does not work without AWS CLI. Setups that use an external OIDC provider are supported with #13804. You can read about it on our blog [^1].

The setup with AWS SSO is supported using the _S3 (Credentials from AWS Command Line Interface) profile_ [^2] connection profile. Configure a bookmark with the field titled _Profile Name_ matching the profile name from ~/.aws/credentials.

You will need to make sure to login using AWS CLI with 
```
aws sso login --profile profilename. 
aws sts get-caller-identity --profile profilename
```

[^1]: https://blog.cyberduck.io/2023/10/18/connect-to-s3-with-web-identity-federation/ 
[^2]: https://docs.cyberduck.io/protocols/s3/#connecting-using-credentials-from-aws-command-line-interface

**Comment by seanturner026** [comment]
To add to the above, dragging this `xml` file into the `Bookmarks` directly was easiest for me. 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>Protocol</key>
	<string>s3</string>
	<key>Provider</key>
	<string>s3-cli</string>
	<key>Nickname</key>
	<string>AWS Identity Center</string>
	<key>Hostname</key>
	<string>s3.amazonaws.com</string>
	<key>Port</key>
	<string>443</string>
	<key>Username</key>
	<string>aws-cli-profile-name-here</string>
	<key>Path</key>
	<string>/</string>
</dict>
</plist>
```
